component/prometheus/interceptor.go

```diff
@@ -0,0 +1,125 @@
+package prometheus
+
+import (
+	"context"
+	"fmt"
+	"sync"
+
+	"github.com/prometheus/prometheus/model/exemplar"
+	"github.com/prometheus/prometheus/model/labels"
+	"github.com/prometheus/prometheus/model/metadata"
+	"github.com/prometheus/prometheus/storage"
+)
+
+type intercept func(ref storage.SeriesRef, l labels.Labels, t int64, v float64) (storage.SeriesRef, labels.Labels, int64, float64, error)
```

Comment:
What do you think about making this more similar to the middleware pattern, meaning:

```go
type intercept func(
  ref storage.SeriesRef, 
  l labels.Labels, 
  t int64, 
  v float64,
  next storage.Appender,
) (storage.SeriesRef, error)
```

This gives the interceptor more control, including skipping the next appender. It also means the `Append` method can be made more minimal: 

```go
// Append satisfies the Appender interface.
 func (a *interceptappender) Append(ref storage.SeriesRef, l labels.Labels, t int64, v float64) (storage.SeriesRef, error) {
 	if ref == 0 {
 		ref = storage.SeriesRef(GlobalRefMapping.GetOrAddGlobalRefID(l))
 	}

        return a.intercept(ref, l, t, v, a.child) 
}
```